examples/gno.land/r/demo/awesomegno/awesomegno.gno

```
@@ -0,0 +1,474 @@
+	"time"
+
+	"gno.land/p/demo/avl"
+	"gno.land/p/demo/avl/rotree"
+	"gno.land/p/demo/seqid"
+	"gno.land/p/demo/ufmt"
+	"gno.land/p/moul/authz"
+	"gno.land/r/leon/hof"
+)
+
+var (
+	dapps     = avl.NewTree()
+	proposals = avl.NewTree()
+
+	nextDappID     seqid.ID
+	nextProposalID seqid.ID
+	nextCategoryID seqid.ID
+
+	categories = avl.NewTree()
+
+	auth = authz.New()
+)
+
+func init() {
+	auth.Current().(*authz.MemberAuthority).AddMember(std.Address("g1ej0qca5ptsw9kfr64ey8jvfy9eacga6mpj2z0y")) // matijamarjanovic
+	hof.Register("AwesomeGno", "Awesome Gno is a registry for gnowesome dApps on gno.land")
+}
+
+func AddAdmin(addr std.Address) error {
+	return auth.Do("add_admin", func() error {
+		if !addr.IsValid() {
+			return ErrInvalidAddress
+		}
+		if err := auth.Current().(*authz.MemberAuthority).AddMember(addr); err != nil {
+			return ErrAdminAlreadyExists
+		}
+		std.Emit("AdminAdded", "address", addr.String())
+		return nil
+	})
+}
+
+func RemoveAdmin(addr std.Address) error {
+	return auth.Do("remove_admin", func() error {
+		if auth.Current().(*authz.MemberAuthority).Tree().Size() <= 1 {
+			return ErrCannotRemoveLastAdmin
+		}
+
+		if err := auth.Current().(*authz.MemberAuthority).RemoveMember(addr); err != nil {
+			return err
+		}
+		std.Emit("AdminRemoved", "address", addr.String())
+		return nil
+	})
+}
+
+func AddCategory(name, description string) (seqid.ID, error) {
+	caller := std.PreviousRealm().Address()
+	if !isAdmin(caller) {
+		return 0, ErrNotAuthorized
+	}
+
+	if name == "" {
+		return 0, ErrCategoryExists
+	}
+
+	sanitizedKey := strings.TrimSpace(name)
+	if categories.Has(sanitizedKey) {
+		return 0, ErrCategoryExists
+	}
+
+	category := Category{
+		ID:          nextCategoryID.Next(),
+		Name:        name,
+		Description: description,
+		CreatedAt:   time.Now(),
+		DappIDs:     []seqid.ID{},
+	}
+
+	categories.Set(sanitizedKey, category)
+
+	std.Emit("CategoryAdded",
+		"id", category.ID.String(),
+		"name", name)
+
+	return category.ID, nil
```

Also, let's work towards more and more top-level functions panicking instead of returning errors. this is a general way to go forward: errors at top level functions should only be used with clear intent, the default should be a panic